accounts-db/src/accounts_index/in_mem_accounts_index.rs

```diff
@@ -737,10 +750,25 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> InMemAccountsIndex<T,
         let entry = map.entry(pubkey);
         m.stop();
         let new_entry_zero_lamports = new_entry.is_zero_lamport();
+        let mut other_slot = None;
         let (found_in_mem, already_existed) = match entry {
             Entry::Occupied(occupied) => {
                 // in cache, so merge into cache
                 let (slot, account_info) = new_entry.into();
+
+                let slot_list = occupied.get().slot_list.read().unwrap();
+
+                // If there is only one entry in the slot list, it means that
+                // the previous entry inserted was a duplicate, which should be
+                // added to the duplicates list too. Note that we only need to do
+                // this for slot_list.len() == 1. For slot_list.len() > 1, the
+                // items, previously inserted into the slot_list, have already
+                // been added. We don't need to add them again.
+                if slot_list.len() == 1 {
```

Comment:
there is a race here. 2 competing insert threads here could both find slot_list.len() = 1. Then, they serilialize access with the write lock on `lock_and_update_slot_list` so that one makes len 2, the next 3. I don't think this is bad in this case, but it is a race condition. `lock_and_update_slot_list` returns slot list len after insertion using write lock. We could set `other_slot` to `None` if `lock_and_update_slot_list` returns `!= 2`, which means someone else won the race.

Reply:
yeah. fixed in ef3e873.